[SR-9615] @escaping and inout should work together but don't

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-9615      |
   |Radar             | None         |
   |Original Reporter | Anandabits (JIRA User)      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, DiagnosticsQoI, Parser, StarterBug        |
   |Assignee          | rayfix (JIRA)      |
   |Priority          | Medium      |

   

   md5: 2869f1f48cc0d67f32ed9775f6516d10

  </details>





**Issue Description:**


The following code does not work, but according to @jckarter it should: <https://twitter.com/jckarter/status/1082421111982219264>

``` java
// 'inout' may only be used on parameters
func foo(f: @escaping inout () -> Void) {}

// @escaping attribute may only be used in function parameter position
func foo(f: inout @escaping () -> Void) {}
```

It must be possible to combine these attributes in order to capture a copy of an inout argument and call it later.

Hello @CodaFi 👋🏻

i'm interested into fixing this issue (as my first contribution to swift), can i get it assigned to me ? 
as a first step i built the compiler, and managed to add support for better diagnostics: 

so now it emits: 

```
/Users/mostafa.essam/Desktop/file.swift:1:30: error: inout expression is implicitly escaping by definition
func m(v: inout @escaping () -> Void) {}
                ~~~~~~~~~~   ^

```
<img width="972" alt="Screenshot 2024-06-09 at 17 39 15" src="https://github.com/apple/swift/assets/2325884/d3ce2130-a3c2-4bde-aeb5-2d9f4d10b8a2">


i'm still evolving the solution, and running the tests to see the affect of my changes.

please tell me what do you think 🙏🏻

--- a/include/swift/AST/DiagnosticsSema.def
+++ b/include/swift/AST/DiagnosticsSema.def
@@ -4015,6 +4015,9 @@ ERROR(autoclosure_function_input_nonunit,none,
 ERROR(escaping_non_function_parameter,none,
       "@escaping attribute may only be used in function parameter position", ())
 
+ERROR(escaping_inout_parameter,none,
+      "inout expression is implicitly escaping", ())
+
 ERROR(escaping_optional_type_argument, none,
       "closure is already escaping in optional type argument", ())
 

--- a/lib/Sema/TypeCheckType.cpp
+++ b/lib/Sema/TypeCheckType.cpp
@@ -3393,6 +3393,10 @@ TypeResolver::resolveAttributedType(TypeRepr *repr, TypeResolutionOptions option
           diagnoseInvalid(repr, repr->getLoc(),
                           diag::escaping_optional_type_argument)
               .fixItRemove(attrRange);
+        } else if (options.is(TypeResolverContext::InoutFunctionInput)) {
+          diagnoseInvalid(repr, repr->getLoc(),
+                          diag::escaping_inout_parameter)
+              .fixItRemove(attrRange);
         } else {
           diagnoseInvalid(repr, loc, diag::escaping_non_function_parameter)
               .fixItRemove(attrRange);

--- a/test/attr/escaping_inout_arugment.swift
+++ b/test/attr/escaping_inout_arugment.swift
@@ -1,0 +1,7 @@
+// RUN: %target-typecheck-verify-swift
+
+let _: (_ v: inout @escaping () -> Void) -> ()
+// expected-error@-1 {{inout expression is implicitly escaping}}{{20-30=}}
+
+func m(v: inout @escaping () -> Void) {}
+// expected-error@-1 {{inout expression is implicitly escaping}}{{17-27=}}

